Use phase functions for EoS in visco_plastic material model

gassmoeller opened the pull request:
This PR builds on #3271 and #3267 and shows how the implemented improvements can be used. In this example it implements simple phase transitions for the equation of state that is used by the visco_plastic material model. This PR in itself is useful, but it is mostly meant as a starting point for @naliboff and @mibillen to start implementing this functionality for the rheology modules they want to use with phase transitions.

jdannberg commented on source/material_model/equation_of_state/interface.cc:
```diff
@@ -36,6 +37,44 @@ namespace aspect
+              phase_in.phase_index = j-c-1;
+              eos_outputs.densities[c] += phase_function.compute_value(phase_in) * (eos_outputs_all_phases.densities[j]-eos_outputs_all_phases.densities[base_index]);
+              eos_outputs.thermal_expansion_coefficients[c] += phase_function.compute_value(phase_in) * (eos_outputs_all_phases.thermal_expansion_coefficients[j]-eos_outputs_all_phases.thermal_expansion_coefficients[base_index]);
+              eos_outputs.specific_heat_capacities[c] += phase_function.compute_value(phase_in) * (eos_outputs_all_phases.specific_heat_capacities[j]-eos_outputs_all_phases.specific_heat_capacities[base_index]);
```
Are you sure this is the correct way to do this? 
Imagine you have two transitions, and the densities are 100, 200, 300. In the highest pressure phase, all phase functions are 1, so you have 100 + (200-100) + (300-100) = 400. Or am I getting this wrong? If I am wrong, that'll at least tell you what documentation to add :smile:

jdannberg commented on source/material_model/visco_plastic.cc:
```diff
@@ -394,13 +395,21 @@ namespace aspect
+
+          MaterialUtilities::PhaseFunctionInputs<dim> phase_inputs(*this,in,i,eos_outputs_all_phases.densities[0],0);
+
+          compute_equation_of_state_phase_transitions(eos_outputs_all_phases,
```
I am not really happy with this name (from the name alone, I find it hard to understand what's happening here). What about something like `phase_average_equation_of_state_outputs`?

jdannberg commented on source/utilities.cc:
```diff
@@ -304,6 +306,27 @@ namespace aspect
+                            ExcMessage("The input parameter " + property_name + " does not have "
+                                       + "the expected number of values for field index " + std::to_string(i) + "."));
+            }
+        }
```
Not that it matters much, but shouldn't this be part of #3271?

jdannberg commented on source/material_model/equation_of_state/multicomponent_incompressible.cc:
```diff
@@ -99,21 +99,29 @@ namespace aspect
         // Retrieve the list of composition names
         const std::vector<std::string> list_of_composition_names = this->introspection().get_composition_names();
 
+        n_phases_per_composition.reset(new std::vector<unsigned int>());
```
Is there a check anywhere that this is consistent with `n_phase_transitions_per_composition`?
If there was a way to always initialize one of these parameters with the other, you wouldn't need the awkward `allow_empty_map` parameter...

gassmoeller commented:
Sorry for taking so long to come back to this. I think I have addressed all remaining comments. I had to adjust the visco_plastic_derivatives tests because they did not correctly set the position for the MaterialModelInputs object so it was impossible to use `get_adiabatic_conditions().density(position)`. Ready from my side.

lhy11009 commented:
Thanks, I will check it out after I finish the work on hand.

On Wed, Feb 12, 2020 at 11:48 AM Rene Gassmoeller <notifications@github.com>
wrote:

> Sorry for taking so long to come back to this. I think I have addressed
> all remaining comments. I had to adjust the visco_plastic_derivatives tests
> because they did not correctly set the position for the MaterialModelInputs
> object so it was impossible to use
> get_adiabatic_conditions().density(position). Ready from my side.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/geodynamics/aspect/pull/3273?email_source=notifications&email_token=AG2WMFPHQ64A24ZBHBM6BYLRCRHBBA5CNFSM4JDWVY7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOELSELPQ#issuecomment-585385406>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AG2WMFIH5XXZHLH4VAAP4W3RCRHBBANCNFSM4JDWVY7A>
> .
>


-- 

Ph.D student of Geophysics
Earth & Planetary Sciences Dept., UC Davis
Davis, CA 95616

jdannberg commented on source/material_model/visco_plastic.cc:
```diff
@@ -394,13 +395,31 @@ namespace aspect
+                                                                   in.pressure[i],
+                                                                   this->get_geometry_model().depth(in.position[i]),
+                                                                   gravity_norm*reference_density,
+                                                                   0);
```
Is that phase index used anywhere before you set the correct values in `phase_average_equation_of_state_outputs`? If not, do you want to set it to NaN?

gassmoeller commented:
Thanks, I think I addressed all comments, simplified the code and added a lot of documentation. Please take a look if this is better.

lhy11009 commented:
Thank you for your comments and changes. It looks like the pull request is
done or nearly done. I'll take a look at it to see what's still needs to be
done.

Stay healthy and have a nice break

On Fri, Mar 20, 2020 at 3:42 PM Rene Gassmoeller <notifications@github.com>
wrote:

> Thanks, I think I addressed all comments, simplified the code and added a
> lot of documentation. Please take a look if this is better.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/geodynamics/aspect/pull/3273#issuecomment-601940980>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AG2WMFNMJGHFWEW6UCINMMTRIPWNTANCNFSM4JDWVY7A>
> .
>


-- 

Ph.D student of Geophysics
Earth & Planetary Sciences Dept., UC Davis
Davis, CA 95616

jdannberg commented on source/material_model/equation_of_state/interface.cc:
```diff
@@ -20,22 +20,68 @@
 
 
 #include <aspect/material_model/equation_of_state/interface.h>
+#include <aspect/simulator_access.h>
```
I guess you checked and you do need this?

gassmoeller replied:
Yes, the explanation is a bit lengthy, but it is because we forward declare `SimulatorAccess` in a header file somewhere else.

jdannberg commented:
Very nice, it's much easier to reach now, thank you!
